File: Algoritmos De Ordenamiento/Algoritmos de ordenamiento Python/Insertion Sort.py
import time
import os
import csv

# Algoritmo de ordenamiento Insertion Sort
def insertion_sort(arr):
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and key < arr[j]:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key
    return arr

# Leer números de un archivo
def read_numbers_from_file(file_path):
    with open(file_path, 'r') as f:
        content = f.read()
        cleaned_content = ''.join(c for c in content if c.isdigit() or c.isspace() or c == '-')
        numbers = list(map(int, cleaned_content.split()))
    return numbers

# Medir el tiempo de ejecución de un algoritmo
def measure_sort_time(sort_function, arr):
    start_time = time.time()
    sort_function(arr.copy()) 
    end_time = time.time()
    return end_time - start_time

# Guardar resultados en un archivo CSV
def save_results_to_csv(results, output_file):
    with open(output_file, 'w', newline='') as csvfile:
        fieldnames = ['Archivo', 'Insertion Sort']
        writer = csv.DictWriter(csvfile, fieldnames=fieldnames)

        writer.writeheader()
        for file, timings in results.items():
            row = {'Archivo': file, 'Insertion Sort': timings['Insertion Sort']}
            writer.writerow(row)

# Función principal
def main():
    folder_path = r'C:\Users\saulj\Downloads\ADA_DATOS-main'  # Ruta de la carpeta que contiene los archivos de datos
    
    results = {}

    # Iterar sobre los archivos
    for filename in os.listdir(folder_path):
        if filename.endswith('.txt'):
            file_path = os.path.join(folder_path, filename)
            numbers = read_numbers_from_file(file_path)
            file_results = {}

            # Medir el tiempo solo de Insertion Sort
            duration = measure_sort_time(insertion_sort, numbers)
            file_results['Insertion Sort'] = duration
            results[filename] = file_results
    
    # Guardar los resultados en un archivo CSV
    save_results_to_csv(results, 'resultados_insertion_sort.csv')

    # Imprimir los resultados
    for file, timings in results.items():
        print(f"Resultados para {file}:")
        print(f"Insertion Sort: {timings['Insertion Sort']:.6f} segundos")
        print()

if __name__ == "__main__":
    main()

File: Algoritmos De Ordenamiento/Algoritmos de Ordenamiento in Java/InsertionSort/src/main/java/com/mycompany/insertionsort/InsertionSort.java
package com.mycompany.insertionsort;

import java.io.*;
import java.util.*;
import java.util.function.Consumer;
public class InsertionSort {

 public static void insertionSort(int[] arr) {
        for (int i = 1; i < arr.length; i++) {
            int key = arr[i];
            int j = i - 1;

            while (j >= 0 && arr[j] > key) {
                arr[j + 1] = arr[j];
                j--;
            }
            arr[j + 1] = key;
        }
    }

    // Swap helper function
    private static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    // Leer números de un archivo
    public static int[] readNumbersFromFile(String filePath) {
        List<Integer> numbers = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
            String line;
            while ((line = br.readLine()) != null) {
                line = line.replaceAll("[\\[\\] ]", ""); // Eliminar corchetes y espacios
                String[] nums = line.split(",");
                for (String num : nums) {
                    numbers.add(Integer.parseInt(num));
                }
            }
        } catch (IOException e) {
            System.err.println("Error al abrir el archivo: " + filePath);
            e.printStackTrace();
        }
        return numbers.stream().mapToInt(i -> i).toArray();
    }

    // Medir el tiempo de ejecución
    public static double measureSortTime(Consumer<int[]> sortFunction, int[] arr) {
        long startTime = System.nanoTime();
        sortFunction.accept(arr);
        long endTime = System.nanoTime();
        return (endTime - startTime) / 1e9; // Convertir a segundos
    }

    public static void main(String[] args) {
        String folderPath = "C:\\Users\\saulj\\Downloads\\ADA_DATOS-main"; // Ruta de la carpeta

        Map<String, Consumer<int[]>> algorithms = new HashMap<>();
        algorithms.put("Insertion Sort", InsertionSort::insertionSort);

        Map<String, Double> results = new HashMap<>();

        // Obtener archivos .txt en el directorio
        File folder = new File(folderPath);
        File[] txtFiles = folder.listFiles((dir, name) -> name.toLowerCase().endsWith(".txt"));

        if (txtFiles != null) {
            for (File file : txtFiles) {
                int[] numbers = readNumbersFromFile(file.getAbsolutePath());

                // Si el archivo tiene demasiados números, saltar
                if (numbers.length > 100000) {
                    System.err.println("Advertencia: El archivo " + file.getName() + " tiene demasiados números.");
                    continue;
                }

                double duration = measureSortTime(algorithms.get("Insertion Sort"), Arrays.copyOf(numbers, numbers.length));
                results.put(file.getName(), duration);
            }
        }

        // Imprimir los resultados en formato de arreglo
        System.out.println("{");
        for (Map.Entry<String, Double> entry : results.entrySet()) {
            System.out.println("  \"" + entry.getKey() + "\": " + entry.getValue() + ",");
        }
        System.out.println("}");
    }


}
